src/optimizer/statistics/stats_storage.cpp

```diff
@@ -6,39 +6,92 @@
+ * cache the columns we use.
+ */
+std::unique_ptr<TableStats> StatsStorage::GetTableStats(const catalog::db_oid_t database_id,
+                                                        const catalog::table_oid_t table_id,
+                                                        catalog::CatalogAccessor *accessor) {
+  auto &stats_storage_value = GetStatsStorageValue(database_id, table_id, accessor);
+  common::SharedLatch::ScopedSharedLatch table_latch{&stats_storage_value.shared_latch_};
+  return stats_storage_value.table_stats_.Copy();
+}
+
+/*
+ * Currently when getting the statistics for a column we get and cache the statistics for the entire table. If we find
+ * that large tables are filling up the cache, we may want to revisit this so we only cache the column requested.
+ */
+std::unique_ptr<ColumnStatsBase> StatsStorage::GetColumnStats(catalog::db_oid_t database_id,
+                                                              catalog::table_oid_t table_id,
+                                                              catalog::col_oid_t column_oid,
+                                                              catalog::CatalogAccessor *accessor) {
+  auto &stats_storage_value = GetStatsStorageValue(database_id, table_id, accessor);
+  common::SharedLatch::ScopedSharedLatch table_latch{&stats_storage_value.shared_latch_};
+  NOISEPAGE_ASSERT(stats_storage_value.table_stats_.HasColumnStats(column_oid), "Should have stats for all columns");
+  return stats_storage_value.table_stats_.GetColumnStats(column_oid)->Copy();
+}
 
-  if (table_it != table_stats_storage_.end()) {
-    OPTIMIZER_LOG_TRACE("There already exists a TableStats object with the given oids.");
-    return false;
+void StatsStorage::MarkStatsStale(catalog::db_oid_t database_id, catalog::table_oid_t table_id,
+                                  const std::vector<catalog::col_oid_t> &col_ids) {
+  StatsStorageKey stats_storage_key{database_id, table_id};
+  auto stats_storage_value = table_stats_storage_.find(stats_storage_key);
+  if (stats_storage_value != table_stats_storage_.end()) {
+    for (const auto &col_id : col_ids) {
+      table_stats_storage_.at(stats_storage_key).table_stats_.GetColumnStats(col_id)->MarkStale();
+    }
   }
-  std::unique_ptr<TableStats> table_stats_ptr = std::make_unique<TableStats>(std::move(table_stats));
-  table_stats_storage_.emplace(stats_storage_key, std::move(table_stats_ptr));
-  return true;
 }
 
-bool StatsStorage::DeleteTableStats(catalog::db_oid_t database_id, catalog::table_oid_t table_id) {
-  StatsStorageKey stats_storage_key = std::make_pair(database_id, table_id);
-  auto table_it = table_stats_storage_.find(stats_storage_key);
+void StatsStorage::InsertTableStats(catalog::db_oid_t database_id, catalog::table_oid_t table_id,
+                                    catalog::CatalogAccessor *accessor) {
+  std::unique_lock<std::mutex> latch(insert_latch_);
```

Comment:
After doing a bit of research, you are right. I naively thought that since we're only inserting, it would be fine to have multiple concurrent readers and one writer, but this is not correct. This will have to be a `SharedLatch`.

Reply:
@malin1993ml Can you take a look at the change I just did and see if it looks right? If so, just resolve this convo.